src/Transient/Internals.hs

```diff
@@ -601,264 +600,260 @@ delData x=  modify $ \st -> st{mfData= M.delete (typeOf x ) (mfData st)}
 forkFinally1 :: IO a -> (Either SomeException a -> IO ()) -> IO ThreadId
-forkFinally1 action and_then =
-  mask $ \restore ->  forkIO $ try (restore action) >>= and_then
-
-free th env= do
-  if isNothing $ parent env
-   then  return Nothing  -- !!>  show th ++ " orphan"
-   else do
-    let msibling= fmap children $ parent env
-
-    case msibling of
-     Nothing -> return Nothing
-     Just sibling  -> do
-       found <- atomically $ do
-                sbs <- readTVar sibling
-                let (sbs', found) = drop [] th  sbs    -- !!> "search "++show th ++ " in " ++ show (map threadId sbs)
-                when found $ writeTVar sibling sbs'    -- !> ("new list",map threadId sbs')
-                return found
-       if (not found && isJust (parent env))
-         then free th $ fromJust $ parent env         -- !!> "toparent"
-         else return $ Just env
-
-   where
-   drop processed th []= (processed,False)
-   drop processed th (ev:evts)| th ==  threadId ev= (processed ++ evts, True)
-                    | otherwise= drop (ev:processed) th evts
-
-hangThread parent child = when(not $ freeTh parent) $ do
-   let headpths= children parent
-   atomically $ do
-       ths <- readTVar headpths
-       writeTVar headpths $  child:ths   -- !!>  "thread added: "++ show (threadId child)
-
--- | kill  all the child threads associated with the continuation context
+forkFinally1 action andThen =
+  mask $ \restore -> forkIO $ try (restore action) >>= andThen
+
+-- | Search for the given ThreadId in the continuation context hierarchy and remove -- it from the children list of an ancestor if possible.
+free :: ThreadId -> EventF -> IO (Maybe EventF)
+free th env =
+  maybe (return Nothing)
+        (\par -> do
+          let sibling = children par
+          found <- atomically $ do
+            sbs <- readTVar sibling
+            let (sbs', found) = drop [] th sbs
+            -- !!> "search "++show th ++ " in " ++ show (map threadId sbs)
+            when found $ writeTVar sibling sbs'
+            -- !> ("new list",map threadId sbs')
+            return found
+          if not found
+          then free th par         -- !!> "toparent"
+          else return $ Just env)
+        (parent env)
+   where drop processed th []  = (processed, False)
+         drop processed th (ev:evts)
+           | th == threadId ev = (processed ++ evts, True)
+           -- WARNING: The order of the siblings is not preserved.
+           -- This shouldn't matter, but if it does, look here.
+           | otherwise         = drop (ev:processed) th evts
+
+-- | If the parent is a not a free thread, add a child context to the parent context
+-- atomically.
+hangThread :: EventF -> EventF -> IO ()
+hangThread parent child =
+  when (not $ freeTh parent) $ do
+    let headpths = children parent
+    atomically $ do
+      ths <- readTVar headpths
+      writeTVar headpths $ child:ths
+    -- !!>  "thread added: "++ show (threadId child)
```

Comment:
Traces again.  There are a lot of such traces which does not occupy separate line.  I've skipped them, but in case of @agocorona will agree all of them can be removed.